drivers/huawei-ups2000.c

```diff
@@ -0,0 +1,616 @@
+	if (crc16_fail) {
+		fatalx(EXIT_FAILURE, "response CRC verification failed!");
+	}
+
+	/* step 5: check response header */
+	if (memcmp(expected_header, ident_response, IDENT_RESPONSE_HEADER_LEN)) {
+		fatalx(EXIT_FAILURE, "unexpected response header!");
+	}
+	ptr += IDENT_RESPONSE_HEADER_LEN;
+
+	/* step 6: extract ident fields */
+	memset(ups2000_ident, 0x00, sizeof(ups2000_ident));
+	for (i = 0; i < UPS2000_IDENT_MAX_FIELDS; i++) {
+		uint8_t type, len;
+
+		if (ptr + 2 > ident_response_end) {
+			break;
+		}
+		type = *ptr++;
+		len = *ptr++;
+
+		if (len + 1 > UPS2000_IDENT_MAX_LEN) {
+			fatalx(EXIT_FAILURE, "response field too long!");
+		}
+		ups2000_ident[i].type = type;
+		ups2000_ident[i].len = len;
+		/*
+		 * Always zero-terminate the bytes, in case the data
+		 * is an ASCII string (i.e. device desc string), libc
+		 * string functions can be used.
+		 */
+		ups2000_ident[i].val[len] = '\0';
+
+		if (ptr + len > ident_response_end) {
+			fatalx(EXIT_FAILURE, "response field too short!");
+		}
+		memcpy(ups2000_ident[i].val, ptr, len);
+		ptr += len;
+	}
+
+	/* step 7: validate device identification field 0x87 and 0x88 */
+	for (i = 0; i < UPS2000_IDENT_MAX_FIELDS; i++) {
+		/* only one device is supported */
+		if (ups2000_ident[i].type == 0x87) {
+			/* so we assume 0x87 must be 1 */
+			ups_count = ups2000_ident[i].val[0] << 24 |
+				    ups2000_ident[i].val[1] << 16 |
+				    ups2000_ident[i].val[2] << 8  |
+				    ups2000_ident[i].val[3];
+		}
+		if (ups2000_ident[i].type == 0x88) {
+			/*
+			 * And only check 0x88, not 0x89, etc. Also copy the
+			 * string for later parsing via strtok().
+			 */
+			ups2000_ident_desc = strdup((char *) ups2000_ident[i].val);
+			break;
+		}
+	}
+	if (ups_count != 1) {
+		fatalx(EXIT_FAILURE, "only 1 UPS is supported, %u found", ups_count);
+	}
+	if (!ups2000_ident_desc) {
+		fatalx(EXIT_FAILURE, "device desc string not found");
+	}
+
+	/* step 8: extract desc fields */
+	for (i = 0; i < UPS2000_DESC_MAX_FIELDS; i++) {
+		char *key, *val;
+		int idx = -1;
+
+		if (i == 0)
+			key = strtok(ups2000_ident_desc, "=");
+		else
+		        key = strtok(NULL, "=");
+		if (!key)
+			break;
+
+		val = strtok(NULL, ";");
+		if (!val)
+			break;
+
+		idx = atoi(key);
+		if (idx + 1 > UPS2000_DESC_MAX_FIELDS || idx < 1) {
+			fatalx(EXIT_FAILURE, "desc index %d is invalid!", idx);
+		}
+		if (strlen(val) + 1 > UPS2000_DESC_MAX_LEN) {
+			fatalx(EXIT_FAILURE, "desc fields %d too long!", idx);
+		}
+		memcpy(ups2000_desc[idx], val, strlen(val) + 1);
+	}
+	free(ups2000_ident_desc);
+
+	/*
+	 * step 9: Validate desc fields that we are going to use are valid.
+	 *
+	 * Note: UPS2000_DESC_DEVICE_ID and UPS2000_DESC_PARALLEL_ID are
+	 * currently unused and unchecked.
+	 */
+	for (int i = UPS2000_DESC_MODEL; i <= UPS2000_DESC_ESN; i++) {
+		if (strlen(ups2000_desc[i]) == 0) {
+			fatalx(EXIT_FAILURE, "desc field %d is missing!", i);
+		}
+	}
+}
+
+
+void upsdrv_initinfo(void)
+{
+	bool in_list;
```

Comment:
Yes, this variable should be initialized. I need to fix it later... It's strange that even with `-Wall -Wextra`, compilers cannot detect a possible use of uninitialized variable... To be fair, I do aware that it's a hard computation problem.

Reply:
Fixed.

Reply:
I can't see the context anymore, but if the originally-uninitialized variable was not read in any codepath before getting assigned, it is deemed okay.

With the "fightwarn" effort I found that most of the time there are no warnings left from clang-8 and older and from gcc-5..gcc-10 (don't have newer locally), so the hordes of warnings to fix come from clang-9, and occasionally from some other set-ups. And then also there are gcc-4.x (and older?) still relevant on systems NUT runs on in practice, whose warnings also matter - but they lately only popped up with introduced uses of GCC pragmas whose syntax support evolved between GCC releases.

Reply:
@jimklimov The context is: it *was* used in the code path. There was a `if (in_list)` check, and it was possible for an uninitialized `in_list` to make the decision, oops, a bug. But strangely, the compiler never issued any warning, not GCC 9 and not clang 10. Thus my comment about correcting issuing warnings for uninitialized variable is a hard computation problem...